### PR TITLE
Adds support for BigCode model family

### PR DESCRIPTION
**Pull request by evellasques**
This PR adds support for training/fine-tuning BigCode family of models (ie.: starcoder, starcoderbase-7b, etc). The multi-query implementation follows the BigCode fork of Megatron (instead of Llama's GQA, they only support a single KV channel and that single channel is replicated across the TP workers).

Below, is a tutorial for anyone willing to use it based on the Code Parrot dataset.

# Code Parrot example

## Getting Started

First download the code parrot dataset
```bash
mkdir -p $HOME/examples_datasets/starcoder
cd $HOME/examples_datasets/starcoder
git clone https://huggingface.co/datasets/codeparrot/codeparrot-clean
cd codeparrot-clean/
git lfs fetch --all && git lfs pull
cd ..
git clone https://huggingface.co/bigcode/starcoder starcoder_model
```

Then from `neuronx-nemo-megatron`:
```bash
python ./nemo/scripts/nlp_language_modeling/preprocess_data_for_megatron.py --preproc-folder \
--input $HOME/examples_datasets/starcoder/codeparrot-clean/ --tokenizer-library huggingface --tokenizer-type \
$HOME/examples_datasets/starcoder/starcoder_model/ --workers 8 --output-prefix \
$HOME/examples_datasets/starcoder/codeparrot-clean-mmap --json-keys=content
```

## Model Compilation
On Slurm:
```bash
sbatch -C trn1.32xlarge --nodes 4 compile.slurm starcoder_7b.sh
```

## Model Training
On Slurm:
```bash
sbatch -C trn1.32xlarge --nodes 4 run.slurm starcoder_7b.sh
```

### Download HF Checkpoint
```bash
cd $HOME/examples_datasets/starcoder
git clone https://huggingface.co/bigcode/starcoderbase-7b
cd starcoderbase-7b && git lfs fetch --all && git lfs pull
```

### Convert HF Checkpoint Format to NeMo
```bash
python ./examples/nlp/language_modeling/checkpoint_conversion/convert_hf_checkpoint_to_nemo_bigcode.py \
--path_to_checkpoint $HOME/examples_datasets/starcoder/starcoderbase-7b/ \
--path_to_tokenizer $HOME/examples_datasets/starcoder/starcoderbase-7b/ \
--config_file $HOME/examples_datasets/starcoder/starcoderbase-7b/config.json \
--output_path $HOME/examples_datasets/starcoder/starcoderbase-7b_nemo_tp_8_pp_2 \
--tp_degree 8 \
--pp_degree 2
```

### Add Reference to NeMo Checkpoint in Training Script
Copy paste the full path of the first shard into `test_bigcode.sh`. Then, configure the `load_xser`, `resume_from_checkpoint` and `use_cpu_initialization` as recommended. It should look like:
```bash
model.resume_from_checkpoint=$HOME/examples_datasets/starcoder/starcoderbase-7b_nemo_tp_8_pp_2/tp_rank_00_pp_rank_000/model_optim_rng.ckpt \
model.use_cpu_initialization=False \
+model.load_xser=True \
```

### Convert NeMo Checkpoint to HF
```bash
python ./examples/nlp/language_modeling/checkpoint_conversion/convert_nemo_checkpoint_to_hf_bigcode.py \
 --path_to_checkpoints ./examples/nlp/language_modeling/nemo_experiments/megatron_bigcode/184/checkpoints/ \
--config_file $HOME/examples_datasets/starcoder/starcoderbase-7b/config.json \
--output_path $HOME/examples_datasets/starcoder/starcoderbase-7b-codeparrot \
--is_xser True
```

### CODE REVIEW

File: nemo/examples/nlp/language_modeling/conf/megatron_bigcode_config.yaml [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+  fp8: False # enables fp8 in TransformerLayer forward
+  fp8_e4m3: False # sets fp8_format = recipe.Format.E4M3 
+  fp8_hybrid: False # sets fp8_format = recipe.Format.HYBRID
+  fp8_margin: 0 # scaling margin 
+  fp8_interval: 1 # scaling update interval
+  fp8_amax_history_len: 1 # Number of steps for which amax history is recorded per tensor
+  fp8_amax_compute_algo: most_recent # 'most_recent' or 'max'. Algorithm for computing amax from history
+  use_emha: False # Use fused multi-head attention for large sequence-length. Note this is not yet supported. Please set to False.
+
+  data:
```
Comment (ptoulme-aws) [inline nemo/examples/nlp/language_modeling/conf/megatron_bigcode_config.yaml:174]: can you add the fine tuning data section of yaml that we in LLama and GPT yamls

File: nemo/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py [hunk 2]
```diff
@@ -321,21 +322,28 @@ def __init__(
             )
 
         super().__init__()
+        self.tokenizer = tokenizer
         self.name = name
         self.indexed_dataset = indexed_dataset
         self.drop_last = drop_last
         self.seq_length = seq_length
+        self.np_rng = np.random.RandomState(seed=seed) # rng state for FIM
```
Comment (ptoulme-aws) [inline nemo/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:330]: is seed same seed as specified in yaml?

File: nemo/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py [hunk 3]
```diff
@@ -394,14 +402,73 @@ def _get_text(self, idx: int) -> np.ndarray:
                 self.indexed_dataset.get(self.doc_idx[doc_index_l], length=offset_l + self.add_extra_token)
             )
             sample = np.concatenate(sample_list)
-        if len(sample) != (self.seq_length + self.add_extra_token):
```
Comment (ptoulme-aws) [inline nemo/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:397]: why are you removing this line

File: nemo/nemo/collections/nlp/modules/common/megatron/bigcode_module.py [hunk 4]
```diff
@@ -0,0 +1,3438 @@
...
+     b: batch size
+     s: sequence length
+     l: number of layers
+    Transformer takes input of size [s, b, h] and returns a
+    tensor of the same size. We use the following arguments:
+        hyperparameters: transformer hyperparameters
+"""
+
+
+class ParallelMLP(MegatronModule, adapter_mixins.AdapterModuleMixin):
```
Comment (ptoulme-aws) [inline nemo/nemo/collections/nlp/modules/common/megatron/bigcode_module.py:104]: I dont understand why this entire file is necessary. This appears to be the entire transformer file duplicated.

File: nemo/examples/nlp/language_modeling/conf/megatron_bigcode_config.yaml (hunk 1 above)
Comment (evellasques) [reply nemo/examples/nlp/language_modeling/conf/megatron_bigcode_config.yaml:174]: Done.

File: nemo/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py (hunk 2 above)
Comment (evellasques) [reply nemo/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:330]: Hey @ptoulme-aws, indeed. Basically I'm extending MegatronGPTModel which passes `cfg.seed` during dataset creation.

https://github.com/aws-neuron/neuronx-nemo-megatron/blob/6038bb80b5ae59d3f4a99ed38a4f68fa0e22665a/nemo/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L876

File: nemo/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py (hunk 3 above)
Comment (evellasques) [reply nemo/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py:397]: Because I now check for that condition in line 451:
```python
elif diff < 0: # too short
 #    sample = np.concatenate([sample, np.full((-1 * diff), self.pad_tok_id)])
            logging.info(
                F' > WARNING: Got sample of length: {len(sample)} for sequence length={self.seq_length+self.add_extra_token}, padding the sample to match sequence length'
            )
            sample = np.array(sample, dtype=np.int64)
            sample = np.pad(
                sample, (0, self.seq_length + self.add_extra_token - len(sample)), mode='constant', constant_values=-1
            )
```

File: nemo/nemo/collections/nlp/modules/common/megatron/bigcode_module.py (hunk 4 above)
Comment (evellasques) [reply nemo/nemo/collections/nlp/modules/common/megatron/bigcode_module.py]: That's a very good remark. My initial goal was to keep it isolated from the GPT-2 implementation. But I put some effort and managed to streamline the code a bit by reusing stuff from the main `module.py` and `transformer.py` that wouldn't affect the BigCode part. Also, I added `BIgCode` to the name of the new classes to highlight the difference. I think that the only other option would be to integrate this directly into the GPT-2 implementation but I think it would add confusion since the way BigCode implemented multi-query attention is not very efficient and I believe GQA is the way to go in the long run.

**Comment by evellasques** [comment]
Any update on this?

**Review by aws-kingrj — APPROVED** [review]

**Review by amithrm — APPROVED** [review]

**Review by musunita — APPROVED** [review]
